File: done/438_find_all_anagrams.java
package done;

/**
 * @author [mst]
 *
 * title or quick description
 *
 * full/optional description: this is a placeholder
 * template for new files creation
 *
 * gains:
 * -[wip] build java template and run tools
[wip] do the same for go
 *
 * @since 2023.02
 * @version 0.1
 **/

////////////////// LIBS
//import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;


////////////////// DECL_IMPL

class Solution {
    // good practice to define constant
    public static final int MAX = 256;

    public static List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        if (p == null || s == null || s.length() == 0 || p.length() > s.length()) {
            return res;
        }

        int[] targetMap = new int[MAX];
        int[] currentMap = new int[MAX];
        // init two maps
        for (int i = 0; i < p.length(); i++) {
            targetMap[p.charAt(i) - 'a']++;
            currentMap[s.charAt(i) - 'a']++;
        }

        int start = 0;
        int end = p.length();
        // Move two pointers
        while(end < s.length()) {
            if (isAnagram(currentMap, targetMap)) {
                res.add(start);
            }
            currentMap[s.charAt(start) - 'a']--;
            start++;
            currentMap[s.charAt(end) - 'a']++;
            end++;
        }
        // Don't forget the last step comparison
        if (isAnagram(currentMap, targetMap)) {
            res.add(start);
        }
        return res;
    }

    private static boolean isAnagram(int[] charMap, int[] targetMap) {
        for (int i = 0; i < MAX; i++) {
            if (charMap[i] != targetMap[i]) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println("[mst] leetcode doodles");

        String s = "cbaebabacd", p = "abc";
		List<Integer> res = findAnagrams(s,p);

        System.out.printf("result for:%s and:%s %n", s,p);
		for (Integer i : res)
			System.out.println(i);
	}
}

File: done/438_find_all_anagrams.cpp
/**
    @author [mst]
    @brief  leetcode problems series
    438_find_all_anagrams
    Given two strings s and p, return an array of all the start indices of p's anagrams in s. You may return the answer in any order.
	An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

	gains:
	-cpp strings/char manipulations
	-sliding window and histogram compination

    features, changelog:
    -[wip] an option to a string/substring based window solution
    -2023.02:	-initial draft

    @version 0.1 2023.02
*/



////////////////// LIBS
#include <iostream>            // usage of console prints
#include <vector>

using namespace std;


////////////////// DECL_IMPL


// build a sliding window for an anagram check at each index of the target string
// this can be considered an expansion of 242_valid_anagram problem
//
// sub: 42T 77S
class Solution {
private:
	static const int MAX = 256;
public:
    vector<int> findAnagrams(string s, string p) {
		vector<int> res;
		int p_size = p.length();
		int s_size = s.length();

		// them basic cases
        if (p.empty() || s.empty() || s_size == 0 || p_size > s_size) {
            return res;
        }

        int targetMap[MAX] = {0};
        int currentMap[MAX] = {0};


		// initial map (sub-pattern sized)
        for (int i = 0; i < p_size; i++) {
            targetMap[p[i] - 'a']++;
            currentMap[s[i] - 'a']++;
        }

		// we now define a comparison window...
        int start = 0;
        int end = p_size;

        while(end < s_size) {
            if (isAnagram(currentMap, targetMap)) {
                res.push_back(start);
            }
			// ... and move it as we go
            currentMap[s[start] - 'a']--;
            start++;
            currentMap[s[end] - 'a']++;
            end++;
        }

        // last case
        if (isAnagram(currentMap, targetMap)) {
            res.push_back(start);
        }

        return res;
    }

	// given the preconditions in the calling function, we diet this
	bool isAnagram(int* charMap, int* targetMap) {
        for (int i = 0; i < MAX; i++) {
            if (charMap[i] != targetMap[i]) {
                return false;
            }
        }
        return true;
    }

};

void printVec (vector<int> v) {
	for (auto x : v) cout << x << " ";
}




////////////////// DRIVER
int main()
{
	Solution sol;
	string s = "cbaebabacd", p = "abc";

	vector<int> res = sol.findAnagrams(s,p);
	cout << "res: ";
	printVec(res);
	cout << endl;

	return 0;
}
